rfcs/20220217-visual_regression_testing.md

```diff
@@ -0,0 +1,226 @@
+
+## Cross platform rendering
+
+Browsers will sometimes render minute differences between platforms, even with the exact same webpage. This means that contributors running Linux might generate different screenshots from those running macOS and again from those running Windows, even if they're all using the same browser and version.
+
+To avoid this, other visual regression testing libraries like BackstopJS support running tests inside a docker container and using the browser inside the container. We can do the same with Playwright (it's just not as easy).
+
+I was about to figure out "a way" of doing it. Here's what I came up with:
+
+1. Create a `./bin/visual-regression.js` file with the following:
+```js
+/**
+ * Run this as a node script so that we can retrieve the `process.cwd`
+ * which isn't possible to evaluate cross-platform inside an package.json script
+ *
+ * We can also automatically sync the local playwright version
+ * with the docker image version. This ensures that the browser
+ * versions expected by the `playwright` binary installed locally
+ * matches the versions pre-installed in the container.
+ */
+const { spawnSync } = require('child_process')
+const fs = require('fs')
+const path = require('path')
+const yaml = require('yaml')
+
+const pnpmLock = yaml.parse(
+  String(fs.readFileSync(path.resolve(process.cwd(), 'pnpm-lock.yaml')))
+)
+
+const playwrightVersion = pnpmLock.devDependencies['@playwright/test']
+
+const args = [
+  'run',
+  '--rm',
+  '-it',
+  '--mount',
+  `type=bind,source=${process.cwd()},target=/src`,
+  '--add-host=host.docker.internal:host-gateway', // This is necessary to make `host.docker.internal` work on Linux. Does it break the command for other OSs?
+  `mcr.microsoft.com/playwright:v${playwrightVersion}-focal`,
+  '/src/bin/visual-regression.sh',
+]
+
+if (process.argv.includes('-u')) {
+  args.push('-u')
+}
+
+spawnSync('docker', args, {
+  stdio: 'inherit',
+})
+2. Create a `./bin/visual-regression.sh` file containing the following (don't forget to `chmod +x` it):
+```sh
+#!/bin/bash
+# This script is meant to run inside the playwright docker container, hence the assumed `/src` directory
+cd /src
+
+# Use npm to avoid having to install pnpm inside the container, it doesn't matter in this case
+npm run test:visual-regression:local -- $1
+```
+3. Add two new package.json scripts:
+```json
+"test:visual-regression": "node ./bin/visual-regression.sh",
+"test:visual-regression:local": "playwright test -c ./test/visual-regression"
+```
+
+All of this together should successfully allow us to run the snapshot tests inside a docker container.
+
+For local debugging purposes, `pnpm test:visual-regression:local --debug` can be used to run the snapshot tests headed in a browser on the host OS. This is useful for debugging the playwright scripts before they actually take a screenshot. Just keep in mind that the final screenshots submitting in a PR need to come from the docker container.
```

Comment:
Can we create the screenshots during the CI action on GitHub, instead of locally in a docker container?
I'm not very comfortable with Docker, and I'm sure many of the community contributors will not be, either. I think it is important not to make Docker installation required for contributing.

Reply:
I guess we could :thinking: That would involve a much bigger process though. I suppose one thing we could do is something like this:

1. You submit a PR
2. GitHub action runs on it to determine if there are snapshot changes
3. If there are snapshot changes, the action creates a new branch based off the PR branch (`<pr-banch>__snapshot-update` or something) that then opens PR targeting your branch with the updated snapshots for review

I'd be happy to explore that :slightly_smiling_face: 

On the other hand, with the `visual-regression.js` script I'm proposing, the interaction with docker should be very minimal, literally just having it installed, you'd have to do nothing else for it to work.

If it really is a hurdle though the supplemental PR with the snapshot updates could do the trick though.

Reply:
Wow, I didn't realize how big a hurdle removing Docker would be. I guess if nothing works, we could checkout the contributors' PRs ourselves and run the visual regression checks.

On Windows Home edition, which is very widespread, you cannot install Docker. You have to set up WSL2 to be able to run it, which can be intimidating.

Reply:
> On Windows Home edition, which is very widespread, you cannot install Docker

:cold_sweat: What?! That's incredible... I'm surprised they let you install WSL if you can't install Docker :confounded:

The GH Action probably wouldn't be too bad to implement, it'd just take some extra work; in particular we'd need to make sure it keeps the supplemental branch up to date and that the process for how to use those PRs is clear, probably via some automated comments left in the parent PR.

I'll sketch this idea out in the RFC so that it's recorded and try to think through some of the edge cases as well. We can decide whether to spend time implementing it later (I'm inclined to say we should, I didn't consider the fact that some folks might not be able to run Docker at all). In any case, it seems like something that would be nice to have eventually anyway.

This actually makes me think we could add a workflow to generate the diff images and upload them as a comment when the snapshots fail. That would be especially helpful and would prevent folks from having to install the pixelmatch library locally (I don't think it's cross platform to Windows).

Reply:
Added notes about this and the proposed solution using supplemental PRs in the RFC here https://github.com/WordPress/openverse/pull/165/commits/eb2b8b54ed3f4256ea8e173ff76dd16a59a366d8

Reply:
> This actually makes me think we could add a workflow to generate the diff images and upload them as a comment when the snapshots fail.

I was thinking we could do something like this! I know the Codecov bot does something similar for code diffs (example here: https://github.com/AllenCellModeling/aicsimageio/pull/377#issuecomment-1046298561). It would be more work though, as you mention.

On the Docker note: while I like the idea of building this so someone doesn't necessarily have to have Docker locally and the checks can still run on GHA, I don't think it's an unreasonable constraint to expect contributors to have Docker installed. Both the catalog and the API stacks function almost exclusively out of containers for local development, so we're already imposing that as a prerequisite in those cases. I don't see a problem extending that to the frontend as well.

Reply:
Shouldn't snapshots be updated consciously on purpose? If they update automatically then in which case those tests would fail?

Reply:
They would only update intentionally and fail the CI on the main PR until they've been updated and committed. It's up to reviewers to ensure that and changes committed to the snapshots are appropriate and match the expected changes from the rest of the PR.